Source/WebKit/UIProcess/Extensions/WebExtension.cpp

```diff
@@ -1548,6 +1557,279 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+        return result;
+    return icon(size);
+}
+
+size_t WebExtension::bestSizeInIconsDictionary(const JSON::Object& iconsObject, size_t idealPixelSize)
+{
+    if (!iconsObject.size())
+        return 0;
+
+#if ENABLE(WK_WEB_EXTENSIONS_ICON_VARIANTS)
+    // Check if the "any" size exists (typically a vector image), and prefer it.
+    if (iconsObject.getValue(anyManifestKey)) {
+        // Return max to ensure it takes precedence over all other sizes.
+        return std::numeric_limits<size_t>::max();
+    }
+#endif
+
+    // Check if the ideal size exists, if so return it.
+    String idealSizeString = String::number(idealPixelSize);
+    if (iconsObject.getValue(idealSizeString))
+        return idealPixelSize;
+
+    // Sort the remaining keys and find the next largest size.
+    Vector<size_t> sizeValues;
+    for (auto key : iconsObject.keys()) {
+        // Filter the values to only include numeric strings representing sizes. This will exclude non-numeric string
+        // values such as "any", "color_schemes", and any other strings that cannot be converted to a positive integer.
+        auto integerValue = parseInteger<size_t>(key);
+        if (integerValue && integerValue > 0)
+            sizeValues.append(*integerValue);
+    }
+
+    if (!sizeValues.size())
+        return 0;
+
+    std::sort(sizeValues.begin(), sizeValues.end());
+
+    size_t bestSize = 0;
+    for (auto size : sizeValues) {
+        bestSize = size;
+        if (bestSize >= idealPixelSize)
+            break;
+    }
+
+    return bestSize;
+}
+
+String WebExtension::pathForBestImageInIconsDictionary(const JSON::Object& iconsObject, size_t idealPixelSize)
+{
+    size_t bestSize = bestSizeInIconsDictionary(iconsObject, idealPixelSize);
+    if (!bestSize)
+        return nullString();
+
+#if ENABLE(WK_WEB_EXTENSIONS_ICON_VARIANTS)
+    if (bestSize == std::numeric_limits<size_t>::max())
+        return iconsObject.getString(anyManifestKey);
+#endif
+
+    return iconsObject.getString(String::number(bestSize));
+}
+
+RefPtr<WebCore::Icon> WebExtension::bestImageForIconsDictionaryManifestKey(const JSON::Object& object, const String& manifestKey, WebCore::FloatSize idealSize, IconsCache& cacheLocation, Error error, const String& customLocalizedDescription)
+{
+    auto currentScales = availableScreenScales();
+    if (auto cachedScales = cacheLocation.getOptional("scales"_s)) {
+        auto scales = std::get_if<Vector<double>>(&*cachedScales);
+        if (!scales || *scales != currentScales)
+            cacheLocation.set("scales"_s, IconVariant(currentScales));
+    } else
+        cacheLocation.set("scales"_s, IconVariant(currentScales));
+
+    auto cacheKey = idealSize.toJSONString();
+    if (auto cacheResult = cacheLocation.getOptional(cacheKey)) {
+        if (std::holds_alternative<RefPtr<WebCore::Icon>>(*cacheResult))
+            return *std::get_if<RefPtr<WebCore::Icon>>(&*cacheResult);
```

Comment:
The old code just did this. It will be `nullptr` if the variant is not an icon, and that is fine.
```suggestion
        return *std::get_if<RefPtr<WebCore::Icon>>(&*cacheResult);
```